### DISCUSSION TITLE
bug error/png.c/MagickPNGErrorHandler/1715

### DISCUSSION THREAD

**Discussion by momomo** [ImageMagick/ImageMagick#4084] [Development]
I am getting an error on one image among a test of 100. 

**identify: Expected 4 bytes; found 0 bytes** 

`/private/var/folders/yd/yyjdq_p91416nc4ktdp50xnm0000gn/T/1629214400008919851-15493196623461502978.png' @ warning/png.c/MagickPNGWarningHandler/1748.
identify: Read Exception `/private/var/folders/yd/yyjdq_p91416nc4ktdp50xnm0000gn/T/1629214400008919851-15493196623461502978.png' @ error/png.c/MagickPNGErrorHandler/1715.

**Command**

magick /../input-16533903733929526133.JPEG -sampling-factor 4:2:0 -quality 85 -interlace JPEG -colorspace RGB -strip -crop 519x506+65+16 /../output-16533903733929526133.m.magick.trim.JPEG

Image attached:

![619359651](https://user-images.githubusercontent.com/9274346/129757500-86f380cd-2050-43e7-9b67-ebbeb942412f.png)

**Comment by fmw42** [comment]
Bad input JPG. Corrupt or created incorrectly.

**Reply by momomo** [reply]
I download this from an external source, I think it is a png but you saying it is a jpg ? 

Did you try ? 

How should I handle this? I would like an output regardless.

**Comment by fmw42** [comment]
The file has a PNG suffix, but your command shows JPG. So did not look carefully at the file suffix. My bad. Nevertheless, I cannot tell what it is since the header is bad.

Exiftool shows a warning about a PNG chunk being truncated. That is likely why it is bad and ImageMagick won't read it that way. Other tools may ignore that warning and try to read it.

exiftool -s -ee -g1 -u -n -D 129757500-86f380cd-2050-43e7-9b67-ebbeb942412f.png
---- ExifTool ----
    - ExifToolVersion                 : 10.51
    - **Warning**                         : **Truncated PNG IEND chunk**
---- System ----
    - FileName                        : 129757500-86f380cd-2050-43e7-9b67-ebbeb942412f.png
    - Directory                       : .
    - FileSize                        : 228703
    - FileModifyDate                  : 2021:08:17 13:09:34-07:00
    - FileAccessDate                  : 2021:08:17 13:11:39-07:00
    - FileInodeChangeDate             : 2021:08:17 13:09:34-07:00
    - FilePermissions                 : 644
---- File ----
    - FileType                        : PNG
    - FileTypeExtension               : PNG
    - MIMEType                        : image/png
---- PNG ----
    0 ImageWidth                      : 403
    4 ImageHeight                     : 257
    8 BitDepth                        : 8
    9 ColorType                       : 2
   10 Compression                     : 0
   11 Filter                          : 0
   12 Interlace                       : 0
    - SRGBRendering                   : 0
    - Gamma                           : 2.2
---- PNG-pHYs ----
    0 PixelsPerUnitX                  : 4724
    4 PixelsPerUnitY                  : 4724
    8 PixelUnits                      : 1
---- Composite ----
    - ImageSize                       : 403x257
    - Megapixels                      : 0.103571

**Comment by momomo** [comment]
Well it is not ideal, since we are not sure how to manage the queue of images to be processed. Do we just fall back to the original, and pretend we did the processing right? Better handling would have been ideal so that such errors do not occur and continue to produce the expected result. I am sure this will happen for far more images. Currently it is only 1 out of ~100. When this increases to 100 000 the issue will grow.

**Comment by snibgo** [comment]
For that file, IM reports an error and won't process it. Exiftool reports a warning. Gimp does not report any error or warning, but thinks the image is entirely black.

So IM is correct in declaring the file to be bad.

Windows Photo Viewer is very tolerant of bad files, and can read the image.

You need to decide how your workflow should deal with bad files. Perhaps the images have great value and you know they are not hostile. Or perhaps they have no value because they have been scooped at random from the internet, so they could be hostile. Perhaps they have been uploaded by your customers, so they might or might not be hostile.

The safest course is to assume, unless you know otherwise, that bad files are hostile and should be rejected.
